### PR TITLE
VMware: Add VMware host inventory plugin

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
docs/docsite/rst/vmware/index.rst
docs/docsite/rst/vmware/vmware_host_inventory.rst
lib/ansible/plugins/inventory/vmware_host_inventory.py

### CODE REVIEW

**Comment by Akasurde** [comment]
@acozine @samccann https://github.com/ansible/ansible/pull/50592 needs to be merged before this.

File: lib/ansible/plugins/inventory/vmware_host_inventory.py [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+    name: vmware_host_inventory
+    plugin_type: inventory
+    short_description: VMware Host inventory source
+    version_added: "2.8"
+    authors:
+        - Abhijeet Kasurde (@Akasurde)
+    description:
+        - Get ESXi host system as inventory hosts from VMware environment.
+        - Uses any file which ends with vmware.yml or vmware.yaml as a YAML configuration file.
+        - The inventory_hostname is always the 'Name' and UUID of the virtual machine. UUID is added as VMware allows ESXi host system with the same name.
```
Comment (pdellaert) [inline lib/ansible/plugins/inventory/vmware_host_inventory.py]: `UUID of the virtual machine`? needs to mention host, i guess

File: lib/ansible/plugins/inventory/vmware_host_inventory.py [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+                               'python or set validate_certs to false in configuration YAML file.')
+
+        ssl_context = None
+        if not self.validate_certs and hasattr(ssl, 'SSLContext'):
+            ssl_context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+            ssl_context.verify_mode = ssl.CERT_NONE
+
+        service_instance = None
+        try:
+            service_instance = connect.SmartConnect(host=self.hostname, user=self.username,
```
Comment (pdellaert) [inline lib/ansible/plugins/inventory/vmware_host_inventory.py]: Instead of managing the SSL context yourself, would it be easier to use `SmartConnectNoSSL` (has been added few years back to PyVmomi).

File: lib/ansible/plugins/inventory/vmware_host_inventory.py [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+        """
+        Get Host System IP Address
+        """
+        ret = None
+        network_system = host_obj.configManager.networkSystem
+        if not network_system:
+            return ret
+
+        for vnic in network_system.networkConfig.vnic:
+            if vnic.portgroup == 'Management Network':
```
Comment (pdellaert) [inline lib/ansible/plugins/inventory/vmware_host_inventory.py]: What if the portgroup of the management vnic is not named "Management Network"? This is not a hardcoded value... I'd just use the first vnic with a valid IP, because there is no guarantee here.

Also, we only support IPv4 addresses, no IPv6?

**Review by pdellaert — CHANGES_REQUESTED** [review]
Good addition, but have a few suggestions (especially on host ip)

File: lib/ansible/plugins/inventory/vmware_host_inventory.py (hunk 3 above)
Comment (Akasurde) [reply lib/ansible/plugins/inventory/vmware_host_inventory.py]: I have changed this method to query current active management nic and return first. I couldn't find any better method than this.

File: lib/ansible/plugins/inventory/vmware_host_inventory.py (hunk 2 above)
Comment (Akasurde) [reply lib/ansible/plugins/inventory/vmware_host_inventory.py]: Not sure about this. I would like to skip this method for now. Might introduce in the next iteration.

File: lib/ansible/plugins/inventory/vmware_host_inventory.py [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+        - Get ESXi host system as inventory hosts from VMware environment.
+        - Uses any file which ends with vmware.yml or vmware.yaml as a YAML configuration file.
+        - The inventory_hostname is always the 'Name' and UUID of the host systems. UUID is added as VMware allows ESXi host system with the same name.
+    extends_documentation_fragment:
+      - inventory_cache
+    requirements:
+      - "Python >= 2.7"
+      - "PyVmomi"
+      - "requests >= 2.3"
+      - "vSphere Automation SDK - For tag feature"
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/vmware_host_inventory.py:28]: This is nice to document the conditional requirements.

File: lib/ansible/plugins/inventory/vmware_host_inventory.py [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+        """Safely get a property or return None"""
+        result = host
+        for attribute in attributes:
+            try:
+                result = getattr(result, attribute)
+            except (AttributeError, IndexError):
+                return None
+        return result
+
+    def _get_managed_objects_properties(self, vim_type, properties=None):
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/vmware_host_inventory.py]: Some of this code looks looks the same in vmware_vm_inventory, might be worth moving into common code?

File: lib/ansible/plugins/inventory/vmware_host_inventory.py [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+        Populate inventory from cache
+        """
+        hostvars = source_data.pop('_meta', {}).get('hostvars', {})
+        for group in source_data:
+            if group == 'all':
+                continue
+            else:
+                self.inventory.add_group(group)
+                self.inventory.add_child('all', group)
+        if not source_data:
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/vmware_host_inventory.py]: If this is true it seems like `hostvars` would also be empty so maybe this block is unneccessary.

File: lib/ansible/plugins/inventory/vmware_host_inventory.py [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+        cacheable_results = self._populate_from_source(source_data, using_current_cache)
+
+        if update_cache:
+            self.cache.set(cache_key, cacheable_results)
+
+    def _populate_from_cache(self, source_data):
+        """
+        Populate inventory from cache
+        """
+        hostvars = source_data.pop('_meta', {}).get('hostvars', {})
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/vmware_host_inventory.py]: Have you tested caching with a file-backed cache plugin? I'm code-reviewing since I can't run this so forgive me if I'm mistaken, but I don't see where you've formatted the cache results with '_meta' and 'hostvars' when caching the data.
If you add
```
cache: True
cache_plugin: jsonfile
cache_connection: /tmp/inventorypath
```
to your config file and use ansible-inventory a couple times you should be able to see if this works as expected.

aws_ec2, which you may have based the caching off, initially formatted results before putting them in the cache. The cacheable results would end up looking something like:
```
{
    'aws_ec2_group_containing_hosts_here': [
        'hostname1',
        'hostname2',
        etc],
    '_meta':
        # hostvars if the host has any associated; hostvars were retrieved by self.inventory.get_host('hostname1')
        'hostvars': {
            'hostname1': {'key1': 'val1', 'key2': 'val2'},
            'hostname2': {'key1': 'val1', 'key2': 'val2'},
            etc
        }
}
```
I did that so the format of the cache from the old ec2.py contrib script and the new plugin looked identical, but a few months ago this was simplified so the plugin just caches the raw data from the API call so the populating from cache or source is the same. I'm not sure if that would apply here well also, but thought I'd mention it.

File: lib/ansible/plugins/inventory/vmware_host_inventory.py (hunk 5 above)
Comment (Akasurde) [reply lib/ansible/plugins/inventory/vmware_host_inventory.py]: OK.

**Comment by Akasurde** [comment]
depends upon https://github.com/ansible/ansible/pull/52642

**Comment by Akasurde** [comment]
@dgmorales Could you please review this ? Thanks.

**Comment by angystardust** [comment]
@Akasurde I'm trying to test your inventory plugin on a 2.7 setup but it gives me the following error:

```
 [WARNING]:  * Failed to parse /home/ansible/ansb27/ansible/vcsa01_hosts.vmware.yml with vmware_host_inventory plugin: _get_managed_objects_properties() got an
unexpected keyword argument 'folder'

  File "/home/ansible/ansb27/lib/python2.7/site-packages/ansible/inventory/manager.py", line 273, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/home/ansible/ansb27/ansible/plugins/inventory/vmware_host_inventory.py", line 191, in parse
    cacheable_results = self._populate_from_source(source_data, using_current_cache)
  File "/home/ansible/ansb27/ansible/plugins/inventory/vmware_host_inventory.py", line 255, in _populate_from_source
    objects = self.pyv._get_managed_objects_properties(vim_type=vim.HostSystem, properties=['name'], folder=root_folder)
```

**Comment by Akasurde** [comment]
> @Akasurde I'm trying to test your inventory plugin on a 2.7 setup but it gives me the following error:

Thanks @angystardust for testing this PR. Did you copy `lib/ansible/plugins/inventory/vmware_vm_inventory.py` as well because with this patch I am editing that file too. 

tl;dr - vmware_vm_inventory.py in this patch contains `_get_managed_objects_properties` API with `folder` as a new parameter.

**Comment by angystardust** [comment]
Yes, @Akasurde I saw that the inventory calls a function contained in `vmware_vm_inventory.py` and copied it in my `inventory` folder. Unfortunately, the vm inventory works as expected but the host inventory returns the error above

**Comment by Akasurde** [comment]
Strange, I am not able to reproduce this on my end. Do you have pycache or pyc files created next to plugin file ? If yes, try deleting them.

**Comment by Akasurde** [comment]
@angystardust Can you ping me on #ansible-devel channel hosted on irc.freenode.net (my irc is akasurde)

**Comment by angystardust** [comment]
@Akasurde sorry for the noise: pulling also the patched `vmware_vm_inventory.py`, it works great! Thanks for your work and for the support :+1:

File: test/integration/targets/inventory_vmware_host_inventory/ansible.cfg [hunk 8]
```diff
@@ -0,0 +1,8 @@
+[defaults]
+inventory = test-config.vmware.yaml
```
Comment (dgmorales) [inline test/integration/targets/inventory_vmware_host_inventory/ansible.cfg:2]: This seems unnecessary to me, as you are always passing with -i on runme.sh anyway.
```suggestion
```

File: test/integration/targets/inventory_vmware_host_inventory/runme.sh [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+ansible-inventory -i ${VMWARE_CONFIG} --list --yaml
+
+# Install TOML for --toml
+${PYTHON} -m pip install toml
+
+# Get inventory using TOML
+ansible-inventory -i ${VMWARE_CONFIG} --list --toml
+
+echo "Check if cache is working for inventory plugin"
+ls "$(pwd)/inventory_cache/vmware_host_*" > /dev/null 2>&1
```
Comment (dgmorales) [inline test/integration/targets/inventory_vmware_host_inventory/runme.sh:63]: The wildcard inside the quote is interpreted literally, leading to ls returning an error and the cleanup trap being called early. See [my changes](https://github.com/ansible/ansible/pull/52782/files#diff-fe5b8e95200b4b3271e8670c972dacfbL62) on an identical code on PR #52782 .
```suggestion
ls "$(pwd)/inventory_cache/vmware_vm_"* > /dev/null 2>&1
```

File: test/integration/targets/inventory_vmware_host_inventory/test_vmware_host_inventory.yml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    - name: store the vcenter container ip
+      set_fact:
+        vcsim: "{{ lookup('env', 'vcenter_host') }}"
+
+    - debug:
+        var: vcsim
+
+    - name: Check that there are 'all' groups present in inventory
+      assert:
+        that: "'{{ item }}' in {{ groups.keys() | list }}"
```
Comment (dgmorales) [inline test/integration/targets/inventory_vmware_host_inventory/test_vmware_host_inventory.yml:18]: This fails on python 2.7, but this code is never ran because of the ls problem mentioned above. See [my changes](https://github.com/ansible/ansible/pull/52782/files#diff-d6070f65d91ac9d1b70aacdeea5828a7R15) on an identical code on PR #52782 .
```suggestion
        that: "item in (groups.keys()|list)"
```

**Review by dgmorales — CHANGES_REQUESTED** [review]
Spotted a few errors I found in identical code on my PR #52782. I'm actually on vacation now 😬, so I think I won't be able to test this on a real vCenter environment before 2 weeks. 

I also saw the changes to vmware_vm_inventory that intersect with the mentioned PR, but I'll comment about them there.

**Review by samccann — APPROVED** [review]
docs portion LGTM
